app/mbm/static/js/userlocations.js

```diff
@@ -0,0 +1,134 @@
+    constructor(app) {
+        this.app = app
+        this.map = app.map
+
+        this.form = null
+        this.addLocationPopup = this.createLocationPopup()
+
+        // Disable double click zooming so we can use that gesture for adding a new location
+        this.map.doubleClickZoom.disable()
+
+        // Store the bound handler so we can remove it during unmount
+        this.dblclickHandler = this.showAddLocationPopup.bind(this)
+        this.map.on('dblclick', this.dblclickHandler)
+
+        // Load user locations from localstorage
+        this.locations = getUserLocations() || {}
+
+        // Render user locations
+        for (const [name, location] of Object.entries(this.locations)) {
+            this.renderSavedLocation(name, location)
+        }
+    }
+
+    createLocationPopup() {
+        const popup = L.popup()
+        const $form = $(`
+            <form>
+                <h4>Add new saved location?</h4>
+                <label for="locationName">Name</label>
+            </form>`
+        )
+        const $nameInput = $('<input id="locationName" name="locationName">')
+        this.$nameInput = $nameInput
+        const $saveButton = $('<button class="btn btn-primary btn-block">Save location</button>')
+        $form.append($nameInput).append($saveButton)
+        $form.submit(this.handleFormSubmission.bind(this))
+
+        this.form = $form[0]
+        popup.setContent(this.form)
+        return popup
+    }
+
+    handleFormSubmission(event) {
+        event.preventDefault()
+
+        const name = this.$nameInput.val()
+        const latlng = this.addLocationPopup.getLatLng()
+
+        this.addLocation(name, latlng.lat, latlng.lng)
+        this.form.reset()
+        this.closeAddLocationPopup()
+    }
+
+    // Create a popup with a form for users to save a new location to the map
+    showAddLocationPopup(clickEvent) {
+        this.addLocationPopup
+            .setLatLng(clickEvent.latlng)
+            .openOn(this.map)
+    }
+
+    // Close the new location popup
+    closeAddLocationPopup() {
+        this.map.closePopup(this.addLocationPopup)
+    }
+
+    // Add a new location to the map and update the locations in localstorage
+    addLocation(name, lat, lng) {
+        const location = { lat, lng }
+        this.locations[name] = location
+        this.renderSavedLocation(name, location)
+        this.saveLocations()
+    }
+
+    // Save user locations to local storage
+    saveLocations() {
+        saveUserLocations(this.locations)
+    }
+
+    // "render" a location to the map by creating a marker at that location,
+    // with a popup containing controls to navigate to/from that marker or to
+    // delete it
+    renderSavedLocation(name, location) {
+        const marker = L.marker([location.lat, location.lng], {
+            title: name,
+            // Make sure these markers sit above the regular source/target markers
+            zIndexOffset: 100,
+        })
```

Comment:
**[Nitpick, non-blocking]** I think it'd be a relatively easy improvement to use a [custom icon](https://leafletjs.com/examples/custom-icons/) like a heart emoji (❤️) to differentiate saved locations from sources/targets; I don't think it needs to block this merge but it'd be nice to do as a fast follow!

Reply:
![image](https://user-images.githubusercontent.com/473542/186011793-807d3d59-e2c5-49b8-9a02-ebb42c335c0f.png)

This was easy and it does look better, nice call!